src/proto/batch/mod.rs

```diff
@@ -0,0 +1,338 @@
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(setter(custom))]
+    pub(crate) complete: Option<Job>,
+}
+
+impl Batch {
+    /// Create a new `BatchBuilder`.
+    pub fn builder(description: impl Into<Option<String>>) -> BatchBuilder {
+        BatchBuilder::new(description)
+    }
+}
+
+impl BatchBuilder {
+    fn build(&mut self) -> Batch {
+        self.try_build()
+            .expect("All required fields have been set.")
+    }
+
+    /// Create a new `BatchBuilder` with optional description of the batch.
+    pub fn new(description: impl Into<Option<String>>) -> BatchBuilder {
+        BatchBuilder {
+            description: Some(description.into()),
+            ..Self::create_empty()
+        }
+    }
+
+    fn success(&mut self, success_cb: impl Into<Option<Job>>) -> &mut Self {
+        self.success = Some(success_cb.into());
+        self
+    }
+
+    fn complete(&mut self, complete_cb: impl Into<Option<Job>>) -> &mut Self {
+        self.complete = Some(complete_cb.into());
+        self
+    }
+
+    /// Create a `Batch` with only `success` callback specified.
+    pub fn with_success_callback(&mut self, success_cb: Job) -> Batch {
+        self.success(success_cb);
+        self.complete(None);
+        self.build()
+    }
+
+    /// Create a `Batch` with only `complete` callback specified.
+    pub fn with_complete_callback(&mut self, complete_cb: Job) -> Batch {
+        self.complete(complete_cb);
+        self.success(None);
+        self.build()
+    }
+
+    /// Create a `Batch` with both `success` and `complete` callbacks specified.
+    pub fn with_callbacks(&mut self, success_cb: Job, complete_cb: Job) -> Batch {
+        self.success(success_cb);
+        self.complete(complete_cb);
+        self.build()
+    }
+}
+
+pub struct BatchHandle<'a, S: Read + Write> {
+    bid: String,
+    prod: &'a mut Producer<S>,
+}
+
+impl<'a, S: Read + Write> BatchHandle<'a, S> {
+    /// ID issued by the Faktory server to this batch.
+    pub fn id(&self) -> &str {
+        self.bid.as_ref()
+    }
+
+    pub(crate) fn new(bid: String, prod: &mut Producer<S>) -> BatchHandle<'_, S> {
+        BatchHandle { bid, prod }
+    }
+
+    /// Add the given job to the batch.
+    pub fn add(&mut self, mut job: Job) -> Result<(), Error> {
+        job.custom.insert("bid".into(), self.bid.clone().into());
+        self.prod.enqueue(job)
+    }
+
+    /// Initiate a child batch of jobs.
+    pub fn start_batch(&mut self, mut batch: Batch) -> Result<BatchHandle<'_, S>, Error> {
+        batch.parent_bid = Some(self.bid.clone());
+        self.prod.start_batch(batch)
+    }
+
+    /// Commit this batch.
+    ///
+    /// The Faktory server will not queue any callbacks, unless the batch is committed.
+    /// Committing an empty batch will make the server queue the callback(s) right away.
+    /// Once committed, the batch can still be re-opened with [open_batch](struct.Producer.html#method.open_batch),
```

Comment:
```suggestion
    /// Once committed, the batch can still be re-opened with [open_batch](Producer::open_batch),
```
I believe this syntax (i.e., just naming the method with Rust naming) will also work.

Reply:
It works just fine! I also forgot to make `BatchHandle` public.